### ISSUE TITLE
Value cannot be null on the DefaultPersistedGrantService after migrating from 4 to 6

### ISSUE THREAD

**Issue by dehaaneric** [DuendeSoftware/Support#214]
**Which version of Duende IdentityServer are you using?**
6.1.5

**Which version of .NET are you using?**
.NET 6 (6.0.9)

**Describe the bug**
Following [these steps](https://docs.duendesoftware.com/identityserver/v6/upgrades/is4_v4_to_dis_v6/) I migrated our IS4 application to Duende IS6. When I run the app, login and request my Grants I receive an empty list of Grants and this in the error log:

```
info: Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker[3]
      Route matched with {action = "Index", controller = "Grants"}. Executing controller action with signature System.Threading.Tasks.Task`1[Microsoft.AspNetCore.Mvc.IActionResult] Index() on controller SocsAuth.Idsrv.Controllers.GrantsController (SocsAuth.Idsrv).
info: Microsoft.EntityFrameworkCore.Infrastructure[10403]
      Entity Framework Core 6.0.9 initialized 'PersistedGrantDbContext' using provider 'Microsoft.EntityFrameworkCore.SqlServer:6.0.9' with options: MigrationsAssembly=SocsAuth.Idsrv
info: Microsoft.EntityFrameworkCore.Database.Command[20101]
      Executed DbCommand (121ms) [Parameters=[@__filter_SubjectId_0='?' (Size = 200)], CommandType='Text', CommandTimeout='30']
      SELECT [p].[Id], [p].[ClientId], [p].[ConsumedTime], [p].[CreationTime], [p].[Data], [p].[Description], [p].[Expiration], [p].[Key], [p].[SessionId], [p].[SubjectId], [p].[Type]
      FROM [PersistedGrants] AS [p]
      WHERE [p].[SubjectId] = @__filter_SubjectId_0
fail: Duende.IdentityServer.Services.DefaultPersistedGrantService[0]
      Failed processing results from grant store.
      System.ArgumentNullException: Value cannot be null. (Parameter 'first')
         at System.Linq.ThrowHelper.ThrowArgumentNullException(ExceptionArgument argument)
         at System.Linq.Enumerable.Union[TSource](IEnumerable`1 first, IEnumerable`1 second, IEqualityComparer`1 comparer)
         at Duende.IdentityServer.Services.DefaultPersistedGrantService.Join(IEnumerable`1 first, IEnumerable`1 second) in /_/src/IdentityServer/Services/Default/DefaultPersistedGrantService.cs:line 119
         at Duende.IdentityServer.Services.DefaultPersistedGrantService.GetAllGrantsAsync(String subjectId) in /_/src/IdentityServer/Services/Default/DefaultPersistedGrantService.cs:line 99
```

I noticed that in the table `PersistedGrants` my `SubjectId` that I'm using has multiple persisted grants. Is that causing the issue? Because after keeping only the latest Persisted Grant, the issue is resolved.

Print-screen on the grants for `SubjectId:42` :
![image](https://user-images.githubusercontent.com/4486709/190454173-c32cdf27-1529-468e-8a1a-2936bdd1d40d.png)

What is causing this issue? Many thanks in advance!

**Comment by dehaaneric** [comment]
I went further with my investigations. Added the Duende IS6 projects to my solution and started debugging. I noticed that he crashes on:

```
    private IEnumerable<Grant> Join(IEnumerable<Grant> first, IEnumerable<Grant> second)
    {
        var list = first.ToList();

        foreach(var other in second)
        {
            var match = list.FirstOrDefault(x => x.ClientId == other.ClientId);
            if (match != null)
            {
                match.Scopes = match.Scopes.Union(other.Scopes).Distinct();
```

`ClientId` and `Scopes` are `null`. As seen in this print-screen:

![image](https://user-images.githubusercontent.com/4486709/190483421-13ef53ba-ab2f-4c4e-bcaf-44ecf33033f8.png)

The `AccessToken` property within the JSON value of the 'Data' column does have these values, but the way its de-serialized doesnt pull them up.

**Comment by brockallen** [comment]
Ok, thanks for that -- it'll be good info as we dig into it.

**Comment by brockallen** [comment]
The odd thing here is that the DefaultPersistedGrantService  has not changed since v4. So something about your data must have changed. Is there anything you can let us know about that?

**Comment by dehaaneric** [comment]
Hi Brock, thanks for taking the time checking this out. This is the content of the Data column of the most recent persistent grant I have in the IS4 database now (some names adjusted to AppName):

```
{
    "CreationTime": "2022-09-19T13:40:49Z",
    "Lifetime": 1296000,
    "ConsumedTime": null,
    "AccessToken": {
        "AllowedSigningAlgorithms": [],
        "Confirmation": null,
        "Audiences": [
            "AppNameAudiencei"
        ],
        "Issuer": "https://login.appname.eu",
        "CreationTime": "2022-09-19T13:40:49Z",
        "Lifetime": 3600,
        "Type": "access_token",
        "ClientId": "io.appname.backend",
        "AccessTokenType": 1,
        "Description": null,
        "Claims": [
            {
                "Type": "client_id",
                "Value": "io.appname.backend",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "scope",
                "Value": "openid",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "scope",
                "Value": "AppScope",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "scope",
                "Value": "offline_access",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "sub",
                "Value": "42",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "auth_time",
                "Value": "1663592110",
                "ValueType": "http://www.w3.org/2001/XMLSchema#integer64"
            },
            {
                "Type": "idp",
                "Value": "local",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "amr",
                "Value": "pwd",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "uid",
                "Value": "41",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "sid",
                "Value": "F2C66750B8518BED24CFB27EC9033164",
                "ValueType": "http://www.w3.org/2001/XMLSchema#string"
            },
            {
                "Type": "iat",
                "Value": "1663594849",
                "ValueType": "http://www.w3.org/2001/XMLSchema#integer64"
            }
        ],
        "Version": 4
    },
    "Version": 4
}
```

The query result from which I've taken this data:

```
select top 10 *
from PersistedGrants
order by CreationTime desc
```

![image](https://user-images.githubusercontent.com/4486709/191089837-31982e9a-135c-40a8-be11-733bc9a6c2aa.png)

The current version on Production (and also the data on the Test-environment that I'm migrating):
```
    <PackageReference Include="IdentityServer4" Version="4.1.2" />
    <PackageReference Include="IdentityServer4.EntityFramework" Version="4.1.2" />
```

Many thanks in advance!

**Comment by brockallen** [comment]
Thanks for that. Off the top of my head I can't tell from this what the issue is.

One question -- do you see any errors in your IdentityServer log? It doesn't look like this is data protection related, since you're coming from an older payload that was not protected in the DB. But any logs might help if they show something unusual. 

So if we can't get any info from the logs then we might need to have you debug this from your side. You should be able to pull in the IdentityServer source into your host to debug the lower level and find out why null is getting returned.

**Comment by dehaaneric** [comment]
We don't have any Errors, Warnings or Criticals related to Grants or the /connect/token endpoints on the IS4 Production or Testing environments. The current IS4 grants implementation works for us perfectly on many iOS and Android devices. Al these clients have good working sliding expiration refresh tokens.

I can work around the `null` error by modifying:

```
var refresh = grants.Where(x => x.Type == IdentityServerConstants.PersistedGrantTypes.RefreshToken)
    .Select(x => _serializer.Deserialize<RefreshToken>(x.Data))
    .Select(x => new Grant
    {
        ClientId = x.ClientId,
        Scopes = x.AuthorizedScopes,
        <...>
    }).ToList();
```

to

```
var refresh = grants.Where(x => x.Type == IdentityServerConstants.PersistedGrantTypes.RefreshToken)
    .Select(x => _serializer.Deserialize<RefreshToken>(x.Data))
    .Select(x => new Grant
    {
        ClientId = x.AccessToken.ClientId,
        Scopes = x.AccessToken.Scopes
        <..>
    }).ToList();
```
because `ClientId` and `Scopes` are not in the JSON-root found in the data column but in the `AccessToken` property.

But I persume this isn't the way to go right?

**Comment by brockallen** [comment]
> because ClientId and Scopes are not in the JSON-root found in the data column but in the AccessToken property.
> But I persume this isn't the way to go right?

Ok, this might be hinting at how we changed the internal AccessToken to AccessTokens structure in the grant and how we dynamically update old records (or possibly fail to do so, as your case might be illustrating). 

I'll do some more looking.

**Comment by dehaaneric** [comment]
As an addition, I extracted the serialization part of the RefreshToken in IS4 and I notice that with these serialization settings:

```
static PersistentGrantSerializer()
{
    _settings = new JsonSerializerSettings
    {
        ContractResolver = new CustomContractResolver()
    };
    _settings.Converters.Add(new ClaimConverter());
    _settings.Converters.Add(new ClaimsPrincipalConverter());
}
```

`Newtonsoft.Json` does not include these `ClientId` and `Scopes` values:

```
        /// <summary>
        /// Gets the client identifier.
        /// </summary>
        /// <value>
        /// The client identifier.
        /// </value>
        public string ClientId => AccessToken.ClientId;

        /// <summary>
        /// Gets the subject identifier.
        /// </summary>
        /// <value>
        /// The subject identifier.
        /// </value>
        public string SubjectId => AccessToken.SubjectId;
```

I tested it with this simple console application:
[NewtonSoftISTests.zip](https://github.com/DuendeSoftware/Support/files/9608240/NewtonSoftISTests.zip)

So I'm cautiously assuming that IS4 does not serialize the properties and their values of `ClientId` and `Scopes` and IS6 does expect them to be there.

**Comment by brockallen** [comment]
Ok, yes -- after doing some more digging, it does look like we're not migrating some of the legacy data the right way. I'm on it.

**Comment by brockallen** [comment]
Ok, I knew with enough digging I'd find it. We do handle this dynamic data upgrade, but at a higher level in the abstraction model:

https://github.com/DuendeSoftware/IdentityServer/blob/50e71c72eb3905acc9fda4e0fa86c2510db665ea/src/IdentityServer/Stores/Default/DefaultRefreshTokenStore.cs#L70-L90

And we even had a test for it:

https://github.com/DuendeSoftware/IdentityServer/blob/ffa4da997b44dcb32103bb26a919fb14cd1cf0f5/test/IdentityServer.UnitTests/Stores/Default/DefaultPersistedGrantStoreTests.cs#L102-L142

But yea, something going right after the `Data` and `IPersistentGrantSerializer` is missing that code to fix up the value. 

Ok, so now just need to find out the best way forward for this. The easiest might just be to duplicate that code in the DefaultPersistedGrantService.

**Comment by brockallen** [comment]
I submitted a PR proposal to fix this. Please have a look.

**Comment by dehaaneric** [comment]
PR looking good imho! Thank you for your quick support.

**Comment by brockallen** [comment]
Sorry for taking so long. We just released patch 6.1.6.
